Update wordpress packages to the latest version

noahtallen opened the pull request:
## Proposed Changes

Now that #77046 has landed, we can update `@wordpress` packages to ~the latest version~ second latest version.

**Note: even if an issue is marked as a resolved, new changes on trunk might still need to be migrated after a rebase.**

Importantly, this _does not_ really impact code running as a normal WordPress script. ETK, for example, uses the dependency extraction plugin, so it is already using the latest APIs provided via the Gutenberg plugin, not via Calypso's dependencies.

This _does_ impact all code using `@wordpress/` packages inside of the Calypso app and other Calypso bundles.

## Testing Instructions
1. Test primary local development work flows, like `yarn` and `yarn start`.
2. Using calypso.live, test major features, like site creation, signup, the editor, etc.
3. If you developed any feature using `@wordpress/` packages, please test that feature!

## Potential high impact areas:
- Global styles. This had to be reworked somewhat.
- WordPress components. The package has seen a lot of changes recently, so anywhere we use it could have a regression.

## To do:

## Complete:
- [x] Test it in the real world!
- [x] Fix remaining client type issues.
- [x] Fix all type errors in packages and ETK.
- [x] Fix type errors in packages blocking yarn install.
- [x] Remove unneeded type packages
- [x] Update resolutions field (I added a new script for doing this!)
- [x] Update all WP packages. (`yarn up '@wordpress/*'`)
- [x] Update yarnrc.yml with peer dependency overrides.
- [x] Fix internal WP components type issue:
```
node_modules/@wordpress/components/build-types/autocomplete/types.d.ts(6,15): error TS2305: Module '"@wordpress/rich-text"' has no exported member 'RichTextValue'.
node_modules/@wordpress/components/build-types/dropdown-menu/types.d.ts(43,24): error TS2339: Property 'role' does not exist on type 'HTMLElement'.
```
This was caused by having custom type definitions in various packages overriding @wordpress/components types. These could be removed now that the package ships its own types.
- [x] ~Update to Jest 29 -- this only impacts ETK, which relies on the WP jest config. This could probably be split out if we want to upgrade the entire repo.~ No longer a blocker after #79097.
- [x] Global styles refactor to work with new package versions: https://github.com/Automattic/wp-calypso/pull/78715
- [x] Fix type issues, especially related to `@wordpress/components` 
  - [x] Anywhere we import a typedef from `@wordpress/components`, it likely needs to be migrated.
  - [x] Anywhere we import a prop type from the package, we need to change to `React.ComponentProps`
  - [x] Anywhere we use `Button` with `isX` props, we need to change to use `variant` instead.
- [x] Fix unit tests.
- [x] Fix issue where a text input was not storing the first input character (https://github.com/Automattic/wp-calypso/pull/79643)

noahtallen committed:
Add script which updates package.json resolutions field

noahtallen committed:
Update @wordpress packages to latest versions in package.json resolutions

noahtallen committed:
Update all @wordpress packages to the latest version

noahtallen committed:
Update internal @wordpress/data peer dependency versions

noahtallen committed:
Resolve most peer dependency issues

noahtallen committed:
Use ComponentProps for some WP component type imports

noahtallen committed:
Rework submenu popover to remove legacy offset props

noahtallen committed:
Fix all Button variant type issues

noahtallen committed:
Remove uneeded button wrapper component in info-tooltip

noahtallen committed:
Fix type for close modal event in page pattern modal

noahtallen committed:
Add ts expect error for ExternalLink required children prop

noahtallen committed:
Add onChange type definition for CustomSelectControl for language-picker

noahtallen committed:
Remove unused @types/wordpress__ packages

noahtallen committed:
Remove uneeded entries in yarnrc.yml

noahtallen committed:
Change another case of isPrimary to variant

noahtallen committed:
Fix global styles package for new WordPress package versions (#78715)

* @automattic/global-styles: Replace @wordpress/edit-site apis with new ones

* Use fixed version of private-apis package to make __private symbol consistent

* Use ExperimentalBlockEditorProvider to initialize Iframe

* Fix unable to get fontFamilies

* Fix styles

* Fix style variation preview

* Fix types

* Replace mergeWith with deepmerge

* Get rid of isEmpty and mapValues in global-stypes pkg

* Add comments

* Revert some changes

* cleanEmptyObject: Fix type

* transformStyles: Fix type

* WP components: Fix types

* Move EditorStyles into the children of the Iframe

---------

Co-authored-by: arthur <arthur.chu@automattic.com>

noahtallen committed:
Delete several @wordpress module declarations that duplicate bundled types

noahtallen committed:
Remove uneeded HTMLElement override

noahtallen committed:
Remove Global Styles wp components type declaration override

noahtallen committed:
Remove manage purchases renew toggle wp components type override

noahtallen commented on packages/global-styles/src/components/global-styles-variations/preview.tsx:
```diff
@@ -133,7 +133,7 @@ const GlobalStylesVariationPreview = ( { title, inlineCss, isFocused, onFocusOut
 											height: normalizedColorSwatchSize * ratio,
 											width: normalizedColorSwatchSize * ratio,
-											background: color,
+											background: color.color,
```
@arthur791004, The types here looked a bit off -- is this what the code is meant to do? It seems that `style` wants the `background` to be a different type, but the color object has three properties.

arthur791004 replied:
The type definition is incorrect here as the `highlightedColors` is `Color[]`. Hence, the type should be

```diff
- ( { slug, color }: { slug: string; color: Color }, index: number ) => (
+ ( { slug, color }: { slug: string; color: string }, index: number ) => (
```

Besides, it might be better to rename the type definition from `Color` to something like `ColorPalette` since the `Color` might be too general and make people misunderstand the meaning 😅

noahtallen replied:
fixed

noahtallen committed:
Add back html element hotfix

noahtallen committed:
Remove imports of deleted type patch files

noahtallen committed:
Fix type errors related to re-declaring Button prop types

noahtallen committed:
Fix type errors related to ExternalLink requiring children

noahtallen committed:
Fix type errors related to Modal isOpen prop

noahtallen committed:
Remove registerPlugin override

noahtallen committed:
Remove target blank from ExternalLink, because it isn't used by the component internally

noahtallen committed:
Fix several prop issues related to ExternalLink (remove target, include children)

noahtallen commented:
Progress for the day:

- All type errors in packages are resolved now! This means we can use `yarn install`.
- All type errors in ETK are resolved.
- I merged in the fix for global styles, so we can actually build the app now :)
- Made good progress on various client type errors, especially the button variant and ExternalLink issues.

Up next, there are about 50 remaining type errors under client, and we'll need to tackle the unit test failures and e2e test suite issue.

Beyond that, the jest update will hopefully resolve the issue with ETK tests, as well as the remaining peer dependency warning.

noahtallen committed:
Add missing @types/wordpress__block-library package

noahtallen committed:
Fix type errors related to Modal open prop

noahtallen committed:
Remove size prop in a couple cases because it doesn't do anything

noahtallen committed:
Fix type errors related to importing component prop types

noahtallen committed:
Fix useState types for date time picker example

tyxla commented:
For the protocol, I'm holding off reviewing this one and iterating on it while we're working on some of its requirements (like #78786 and its own dependency #79097)

noahtallen committed:
yarn dedupe

noahtallen committed:
Fix type issue with NavigatorButton wrapper class name

noahtallen committed:
Remove another wp components type override file

noahtallen committed:
Update onNavigatorPathChange argument which is sometimes undefined

noahtallen committed:
Fix a type issue with global styles context

noahtallen committed:
Fix eslint issues with hooks in modified files

noahtallen committed:
Remove unused translate import

noahtallen commented:
> For the protocol, I'm holding off reviewing this one and iterating on it while we're working on some of its requirements (like https://github.com/Automattic/wp-calypso/pull/78786 and its own dependency https://github.com/Automattic/wp-calypso/pull/79097)

@tyxla I just rebased on trunk after merging #79097. I think the Jest PR should no longer be a blocker, since we no longer have a dependency on `@wordpress/scripts`. (this means we don't have a peer dependency conflict any more.)

I also fixed up some more type issues that got introduced on trunk which brings us to a nice milestone! `yarn install` completes with no warnings or errors :D

tyxla commented:
> > For the protocol, I'm holding off reviewing this one and iterating on it while we're working on some of its requirements (like #78786 and its own dependency #79097)
> 
> @tyxla I just rebased on trunk after merging #79097. I think the Jest PR should no longer be a blocker, since we no longer have a dependency on `@wordpress/scripts`. (this means we don't have a peer dependency conflict any more.)
> 
> I also fixed up some more type issues that got introduced on trunk which brings us to a nice milestone! `yarn install` completes with no warnings or errors :D

This is amazing, thank you @noahtallen 🙌 

I'm hoping to be able to help with the failures throughout the week.

noahtallen committed:
Fix type issue where reader button couldn't be disabled

noahtallen committed:
Remove disabled from a Button instance which is a link

noahtallen committed:
Fix state type related to FontSizePicker

noahtallen committed:
Fix odd issue with query control types

noahtallen committed:
Fix issue with SlotFill not being typed

noahtallen committed:
Remove broken types from ToolbarGroup as it's just an example

noahtallen committed:
Fix issue with arr.includes undefined

noahtallen committed:
Fix issue with combobox onchange arg type

noahtallen committed:
Remove disabled from another button link

noahtallen committed:
Cast translateresult to string for input to label

noahtallen committed:
Cast string to more specific type for log tab

noahtallen committed:
Fix onChange type for another combobox

noahtallen committed:
Add missing onhover prop to component

noahtallen committed:
Fix issue with Token value types

noahtallen committed:
Add missing labels to notification toggles

noahtallen committed:
Remove useless wrapper component

noahtallen committed:
Add types to callback args

noahtallen committed:
Remove uneeded ts expect err and update component props for combobox control

noahtallen committed:
fix color type definition for global styles

noahtallen committed:
Remove uneeded types in callback

noahtallen committed:
Remove useless component wrapper

noahtallen commented:
I just worked through type issues after rebasing again. I think we should be down to only around 15 or so!

tyxla committed:
Utilize useLocalizeUrl() where necessary

tyxla committed:
Fix some test snapshots

tyxla committed:
Fix composite checkout tests by introducing field reset

tyxla committed:
dedupe

tyxla committed:
Fix most of the remaining TS errors

tyxla commented:
I've made some progress on this one today:

* Rebased a few times to resolve conflicts
* Added a few `localizeUrl` calls where necessary (in files we touched) to address ESLint errors.
* Fixed the failing unit tests.
* Fixed most of the TS errors - only 2 remain now.

Unfortunately I have to wrap up for the day, but feel free to continue @noahtallen 🙌

noahtallen committed:
Remove uneeded wrapper components

noahtallen committed:
Add expect-error to incorrect .then error

sirbrillig commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/upgrade-modal.tsx:
```diff
@@ -112,7 +112,7 @@ const UpgradeModal = ( { slug, isOpen, closeModal, checkout }: UpgradeModalProps
 						{
 							components: {
-								link: <ExternalLink target="_blank" href="https://woocommerce.com/" />,
+								link: <ExternalLink children={ null } href="https://woocommerce.com/" />,
```
I haven't tested this but I'm curious: I don't know how `translate()` component interpolation works but will `children` be correctly overridden when this is put into the string? I see this pattern has been used in a lot of places so I guess so but I was wondering if you'd tried it.

noahtallen replied:
We definitely need to test it out, but I did take a look at the i18n code and I'm _pretty_ sure it handles this scenario!

https://github.com/Automattic/wp-calypso/blob/54ff4556c2e4eeae2810d0e93ae212cb4e3b0731/packages/interpolate-components/src/index.js#L75-L83

noahtallen replied:
I've verified it works -- it does correctly interpolate the string into the component.

sirbrillig commented on client/my-sites/checkout/composite-checkout/payment-methods/credit-card/assign-to-all-payment-methods.tsx:
```diff
@@ -37,19 +37,23 @@ export default function AssignToAllPaymentMethods( {
+							},
+						}
+						// As far as I can tell, label will correctly render the
+						// component, so we cast to string to make the types work.
```
We might want to open an issue on `@wordpress/components` to update the type of `CheckboxControl` so that it accepts something more broad for `label`.

noahtallen replied:
Good idea! I have this on my list of things to look into in Gutenberg

sirbrillig commented on client/my-sites/checkout/composite-checkout/test/credit-card.tsx:
```diff
@@ -113,6 +122,9 @@ describe( 'Credit card payment method', () => {
 		} );
+
+		// Manually reset the `wpcom-credit-card` store fields.
+		render( <ResetCreditCardStoreFields /> );
```
Interesting! This is fine. I'd be curious what triggered the need for this (presumably the tests were leaking).

sirbrillig commented on client/landing/subscriptions/components/site-subscription-page/site-subscription-details.tsx:
```diff
@@ -256,9 +256,8 @@ const SiteSubscriptionDetails = ( {
-								isPrimary
+								variant="primary"
 								href="/me/purchases"
-								disabled={ unsubscribing }
```
Why was this `disabled` removed? The others in this file remain...

noahtallen replied:
When `Button` is given an `href`, it renders an `a`. As a result, `disabled` isn't permitted via types when the button is actually a link. I'm not sure what has functionally changed inside the component, but the solution was either to remove the style (disabled) or to remove the functionality (href), so I figured I'd keep the link.

(This applies everywhere I moved away from `disabled` for buttons.)

sirbrillig replied:
Ah! Good point. I hadn't considered that. Thanks for the explanation!

sirbrillig commented on client/components/marketing-survey/cancel-purchase-form/step-components/upsell-step.tsx:
```diff
@@ -46,19 +46,18 @@ function Upsell( { image, ...props }: UpsellProps ) {
 							props.onAccept?.();
 						} }
 						isBusy={ busyButton === 'accept' }
-						disabled={ Boolean( busyButton && busyButton !== 'accept' ) }
```
Why was this `disabled` removed?

sirbrillig reviewed (commented):
I don't see anything particularly blocking from a Payments perspective. I left a few questions. Thanks for working on this!

arthur791004 commented:
Proposed https://github.com/Automattic/wp-calypso/pull/79275 to resolve the broken layout on the Assembler screen.

worldomonation commented:
Change looks fine to me, but it's because it barely affects E2E-related packages.

On that topic, let me know if there's anything I can do to help with this change or testing of it @noahtallen.

arthur791004 committed:
Assembler: Fix the main layout is broken (#79275)

simison commented:
Tested `/setup/newsletter` onboarding flow and all was fine but I noticed disabled button styles change:

**Before**
<img width="464" alt="Screenshot 2023-07-12 at 15 26 24" src="https://github.com/Automattic/wp-calypso/assets/87168/5bee15a3-dd06-4c2b-b54c-e6485337cb33">

**After**
<img width="457" alt="Screenshot 2023-07-12 at 15 26 28" src="https://github.com/Automattic/wp-calypso/assets/87168/1e4508e9-8a87-4f40-9cfb-f82f533a53a6">

Would you prefer we fix in this PR, before this PR, or after? There could be other similar instances around Calypso.

dlind1 commented on packages/language-picker/src/language-picker.tsx:
```diff
@@ -208,7 +212,11 @@ function LanguagePicker< TLanguage extends Language >( {
 					{ languagesToRender.map( ( language ) => (
 						<Button
-							isPrimary={ selectedLanguage && language.langSlug === selectedLanguage.langSlug }
+							variant={
```
The styling seems to be off now, tested in calypso.live.

Before:
<img width="1602" alt="language-picker-before" src="https://github.com/Automattic/wp-calypso/assets/7847633/0cf2a3d9-f952-4d70-8b17-2fa449abbf39">

After:

<img width="1602" alt="language-picker-after" src="https://github.com/Automattic/wp-calypso/assets/7847633/664c0fff-d82a-40bf-90b2-3c0026637fda">

noahtallen replied:
Thanks! I assumed the `secondary` variant of the button was the default, but it's not. I restored the previous styles by changing variant to undefined when it's not primary.

noahtallen committed:
Comment out wrapperClassName which causes a type error

noahtallen committed:
Change a few HStack props from alignItems to alignment

noahtallen commented:
@simison I checked the CSS for the buttons on trunk vs the PR, and I think the issue is the primary button's backgronud uses the variable `--wp-admin-theme-color`, which is changing very slightly from this:

<img width="770" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6265975/b72b5ebd-e711-4f69-a2aa-3ac4cd51899a">

To a more vibrant (almost purpleish) color:

<img width="765" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6265975/709727e9-0da6-4726-aeb5-3efa2613eafc">

The change feels a lot more drastic when the button is disabled, which applies 0.5 opacity!

I'll leave it up to your team to decide on a solution to this! (We might want to set a specific background color for the `NextButton`?)

arthur791004 committed:
Assembler: Fix the styles of the navigator button (#79338)

noahtallen commented:
By the way, I'm planning to merge this update early next week, and I'll do another rebase at that time. In the meantime, feel free to continue testing and providing feedback!

tyxla commented:
With regard to the reported unplanned styling changes, I'd recommend comparing this branch vs trunk on the WordPress components galley page in devdocs: https://wpcalypso.wordpress.com/devdocs/wordpress-components-gallery

If you compare directly, you'll see a few things that might be unexpected. Most is spacing which I'd hazard a guess may be fine, but others could potentially need some additional love, for example:

`trunk`:
![Screenshot 2023-07-14 at 15 03 00](https://github.com/Automattic/wp-calypso/assets/8436925/0993a90e-3b14-4c0a-9c1b-dd8c5397c095)

this branch:

![Screenshot 2023-07-14 at 15 02 45](https://github.com/Automattic/wp-calypso/assets/8436925/11511a08-40e5-4b91-befc-1643b9a3d831)

noahtallen commented:
Ah, the clipboard button is probably a mistake. We removed `isPrimary`, but didn't change it to `variant="primary"`, thus giving it the default style

noahtallen committed:
Lock file update

noahtallen committed:
Make Button variants more consistent with trunk

noahtallen committed:
Use better approach for picking anchor button types

arthur791004 reviewed (commented):
Tested again, and it still works as expected 👍👍👍

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8600733

Hi @noahtallen, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Set staging site.`
- `Disable email notifications`
- `Disable mobile notifications`
- `Enable mobile notifications`
- `Disable SMS notifications`
- `Enable SMS notifications`

Thank you in advance!

yashwin commented:
@noahtallen: I see a label prop has been added to some `ToggleControl` components and this has caused some issues in the Jetpack Pro Dashboard.

<img width="469" alt="Screenshot 2023-07-24 at 11 20 31 AM" src="https://github.com/Automattic/wp-calypso/assets/10586875/89393c3b-516a-49de-a4f0-f1f75102fc6c">

noahtallen commented:
Thanks @yashwin -- I thought the label text was hidden when making the change, so I didn't catch this. thanks for the fix!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8600733).
